services/web/server/src/simcore_service_webserver/garbage_collector_tasks_users.py

```diff
@@ -0,0 +1,112 @@
+"""
+    Scheduled tasks addressing users
+
+"""
+
+import asyncio
+import logging
+from typing import AsyncIterator, Callable
+
+from aiohttp import web
+from aiopg.sa.engine import Engine
+from models_library.users import UserID
+from servicelib.logging_utils import log_context
+from tenacity import retry
+from tenacity.before_sleep import before_sleep_log
+from tenacity.wait import wait_exponential
+
+from ._constants import APP_DB_ENGINE_KEY
+from .login.utils import notify_user_logout
+from .security_api import clean_auth_policy_cache
+from .users_db import update_expired_users
+
+logger = logging.getLogger(__name__)
+
+CleanupContextFunc = Callable[[web.Application], AsyncIterator[None]]
+
+_SEC = 1
+
+_PERIODIC_TASK_NAME = f"{__name__}.update_expired_users_periodically"
+_APP_TASK_KEY = f"{_PERIODIC_TASK_NAME}.task"
+
+
+async def notify_user_logout_all_sessions(
+    app: web.Application, user_id: UserID
+) -> None:
+    #  NOTE kept here for convenience
+    with log_context(
+        logger,
+        logging.INFO,
+        "Forcing logout of %s from all sessions",
+        f"{user_id=}",
+    ):
+        try:
+            await notify_user_logout(app, user_id, client_session_id=None)
+        except Exception:  # pylint: disable=broad-except
+            logger.warning(
+                "Ignored error while notifying logout for %s",
+                f"{user_id=}",
+                exec_info=True,
+            )
+
+
+@retry(
+    wait=wait_exponential(min=5 * _SEC),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+)
+async def _update_expired_users_periodically(
+    app: web.Application, wait_interval_s: float
+):
+    """Periodically check expiration dates and updates user status
+
+    It is resilient, i.e. if update goes wrong, it waits a bit and retries
+    """
+    engine: Engine = app[APP_DB_ENGINE_KEY]
+    assert engine  # nosec
+
+    while True:
+        if updated := await update_expired_users(engine):
+            # expired users might be cached in the auth. If so, any request
+            # with this user-id will get thru producing unexpected side-effects
+            clean_auth_policy_cache(app)
+
+            # broadcast force logout of user_id
+            for user_id in updated:
+                logger.info("User account with %s expired", f"{user_id=}")
+
+                # NOTE: : this notification will never reach sockets because it runs in the GC!!
+                # We need a mechanism to send messages from GC to the webservers
```

Comment:
This seems more like an important `TODO` to me than a note. This means that if a user account is expired he is
- not logged out
- sockets (i.e. dynamic services) are not terminated

Correct?

Reply:
See answer below and new entry in ``docs/coding-conventions.md``